components/script/dom/node.rs

```diff
@@ -546,26 +597,24 @@ impl Node {
             self.inclusive_descendants_version(),
             doc.inclusive_descendants_version(),
         ) + 1;
-        for ancestor in self.inclusive_ancestors() {
+        for ancestor in self.inclusive_ancestors(ShadowIncluding::No) {
             ancestor.inclusive_descendants_version.set(version);
         }
         doc.inclusive_descendants_version.set(version);
     }
 
     pub fn dirty(&self, damage: NodeDamage) {
         self.rev_version();
-        if !self.is_in_doc() {
+        if !self.is_connected() {
             return;
         }
 
         match self.type_id() {
-            NodeTypeId::CharacterData(CharacterDataTypeId::Text) => self
-                .parent_node
-                .get()
-                .unwrap()
-                .downcast::<Element>()
-                .unwrap()
-                .restyle(damage),
+            NodeTypeId::CharacterData(CharacterDataTypeId::Text) => {
+                if let Some(parent) = self.composed_parent_node() {
```

Comment:
This should use instead:

`self.parent_node.get().unwrap().dirty(damage)`,

And this match should handle `ShadowRoot`, deferring to the host. That should prevent the panics you were seeing.